src/raft.rs

```diff
@@ -1851,35 +1840,43 @@ impl<T: Storage> Raft<T> {
     /// Indicates whether state machine can be promoted to leader,
     /// which is true when its own id is in progress list.
     pub fn promotable(&self) -> bool {
-        self.prs().voters().contains_key(&self.id)
+        self.prs().voter_ids().contains(&self.id)
     }
 
-    fn add_voter_or_learner(&mut self, id: u64, is_learner: bool) {
-        if self.prs().voters().contains_key(&id) {
-            if is_learner {
-                info!(
-                    "{} ignored add learner: do not support changing {} from voter to learner",
-                    self.tag, id
-                );
-            }
-            // Ignore redundant add voter.
-            return;
-        } else if self.prs().learners().contains_key(&id) {
-            if is_learner {
-                // Ignore redundant add learner.
+    fn add_voter_or_learner(&mut self, id: u64, learner: bool) {
+        debug!(
+            "Adding node (learner: {}) with ID {} to peers.",
+            learner, id
+        );
+        let progress = Progress::new(self.raft_log.last_index(), self.max_inflight, learner);
```

Comment:
Seems more readable to me to move this to L1867. Otherwise it may be confusing as L1855 overlaps the variable name in its scope. And I think it should be `last_index + 1`.